### PR TITLE
Added modifications to right click menus  

### PR DESCRIPTION
**Pull request by mmiscool**
Added "Zoom  to fit" and "Select all" to right click menu for no selection.

Added "Toggle Construction", "Dimension" to right click menu for selection of line/curve geometry.

Added "Center View At Point" and "Fillet (Tangent arc at point)" to right click menu for selection of point.

Fixed "PNG" file export to export image right side up with out the toolbar in view.

Added menu entry "Save With Exports" command to automatically save a "png", "html" and "stl"
file along side the "slvs" file.

### CODE REVIEW

**Comment by whitequark** [comment]
Most importantly, when submitting a PR, please implement exactly one feature or bugfix per commit. This makes it far easier to review, bisect, and cherry-pick these.

Overall formatting issues with this patch:
  * Use 4 spaces, never tabs.
  * Don't leave trailing whitespace around.
  * Don't just add empty lines around.

File: src/mouse.cpp [hunk 1]
```diff
@@ -589,9 +595,13 @@ void GraphicsWindow::MouseRightUp(double x, double y) {
                 if(r->type == Request::Type::CUBIC) addAfterPoint++;
                 AddContextMenuItem("Add Spline Point", ContextCommand::ADD_SPLINE_PT);
             }
+
         }
 
         if(gs.points == 1) {
+			
+			AddContextMenuItem("Fillet", ContextCommand::TANGENT_ARC);
```
Comment (whitequark) [inline src/mouse.cpp]: This should be called something like "Create Tangent Arc"; we don't call it "Fillet" elsewhere.

File: src/solvespace.cpp [hunk 2]
```diff
@@ -466,6 +466,15 @@ void SolveSpaceUI::MenuFile(Command id) {
             SS.GetFilenameAndSave(/*saveAs=*/false);
             break;
 
+		case Command::SAVE_WITH_EXPORTS:
+			SS.GetFilenameAndSave(/*saveAs=*/false);
+
+			SS.ExportAsPngTo(SS.saveFile + ".png");
+			SS.ExportMeshTo(SS.saveFile + ".html");
+			SS.ExportMeshTo(SS.saveFile + ".stl");
```
Comment (whitequark) [inline src/solvespace.cpp]: I don't think this is generic enough to be merged. This is just a set of arbitrary three formats that are useful to you, and aren't configurable in any way.

File: src/export.cpp [hunk 3]
```diff
@@ -1099,15 +1097,18 @@ void SolveSpaceUI::ExportAsPngTo(const std::string &filename) {
         SS.GW.Paint();
     });
```
Comment (whitequark) [inline src/export.cpp]: The right fix is, I think, adding
```
#else
    SS.GW.Paint();
```
after this line.

File: src/export.cpp [hunk 4]
```diff
@@ -220,8 +220,6 @@ void SolveSpaceUI::ExportViewOrWireframeTo(const std::string &filename, bool exp
             for(Constraint &c : SK.constraint) {
                 c.Draw(Constraint::DrawAs::DEFAULT, &canvas);
             }
-
-            canvas.Clear();
```
Comment (whitequark) [inline src/export.cpp]: Why was this removed? You've brought back a memory leak.

**Comment by mmiscool** [comment]
I think I cleaned up every thing required.

**Comment by whitequark** [comment]
No, not really. Please see https://github.com/solvespace/solvespace/pull/33/files.

**Comment by mmiscool** [comment]
Possibly ok now. . .

File: src/export.cpp [hunk 5]
```diff
@@ -1096,18 +1095,22 @@ void SolveSpaceUI::ExportAsPngTo(const std::string &filename) {
...
     });
+#else
+    SS.GW.Paint();
 #endif
-    SS.showToolbar = prevShowToolbar;
-
     // Somewhat hacky way to invoke glReadPixels without dragging in all OpenGL headers.
     OpenGl1Renderer canvas = {};
     canvas.camera = SS.GW.GetCamera();
+    SS.showToolbar = false;
```
Comment (whitequark) [inline src/export.cpp]: All the other modifications should not be necessary once the `#else` branch is added.

File: src/export.cpp [hunk 6]
```diff
@@ -220,7 +220,6 @@ void SolveSpaceUI::ExportViewOrWireframeTo(const std::string &filename, bool exp
             for(Constraint &c : SK.constraint) {
                 c.Draw(Constraint::DrawAs::DEFAULT, &canvas);
             }
-
```
Comment (whitequark) [inline src/export.cpp]: Stray whitespace.

File: src/mouse.cpp [hunk 7]
```diff
@@ -535,10 +535,12 @@ void GraphicsWindow::MouseRightUp(double x, double y) {
         SS.ScheduleShowTW();
     }
     GroupSelection();
-
+
```
Comment (whitequark) [inline src/mouse.cpp]: Stray whitespace.

File: src/mouse.cpp [hunk 8]
```diff
@@ -535,10 +535,12 @@ void GraphicsWindow::MouseRightUp(double x, double y) {
         SS.ScheduleShowTW();
     }
     GroupSelection();
-
+    
     bool itemsSelected = (gs.n > 0 || gs.constraints > 0);
     int addAfterPoint = -1;
 
+
+
```
Comment (whitequark) [inline src/mouse.cpp]: Stray whitespace.

File: src/mouse.cpp [hunk 9]
```diff
@@ -589,9 +600,13 @@ void GraphicsWindow::MouseRightUp(double x, double y) {
                 if(r->type == Request::Type::CUBIC) addAfterPoint++;
                 AddContextMenuItem("Add Spline Point", ContextCommand::ADD_SPLINE_PT);
             }
+
```
Comment (whitequark) [inline src/mouse.cpp]: Stray whitespace, and so on, and so on.

File: src/ui.h [hunk 10]
```diff
@@ -139,6 +140,12 @@ enum class ContextCommand : uint32_t {
     SNAP_TO_GRID     = 0x140,
     REMOVE_SPLINE_PT = 0x141,
     ADD_SPLINE_PT    = 0x142,
+	CONSTRUCTION     = 0x143,
```
Comment (whitequark) [inline src/ui.h]: Indentation.

File: src/ui.h [hunk 11]
```diff
@@ -17,6 +17,7 @@ enum class Command : uint32_t {
     OPEN_RECENT,
     SAVE,
     SAVE_AS,
+    SAVE_WITH_EXPORTS,
```
Comment (whitequark) [inline src/ui.h]: There is no SAVE_WITH_EXPORTS command any more.

**Comment by whitequark** [comment]
Most importantly, the pull request is a conglomeration of many commits and many features. This is bad. It makes it hard to:
  * bisect the bugs;
  * revert commits, if necessary;
  * cherry-pick commits into another branch, if necessary;
  * explore the history and figure out why was some code written.

There should be one commit per feature, and nothing else.

git has a feature called "rebase" that allows you to rearrange commits into a more descriptive order, although for an amount of code it might be easier to add just the needed changesets from scratch.

**Comment by whitequark** [comment]
I've fixed the PNG export problem you've discovered in master, so you can drop that part, at least.

**Comment by whitequark** [comment]
> Added "Zoom to fit" and "Select all" to right click menu for no selection.

Implemented in master. "Zoom to Fit" is implemented such that it can be used when more than one entity (or constraint) is selected, making it much more useful.

> Added "Toggle Construction"

Implemented in master, again works with many entities.

> "Dimension" to right click menu for selection of line/curve geometry.
> 
> Added "Center View At Point" and "Fillet (Tangent arc at point)" to right click menu for selection of point.

These don't seem obviously useful enough. In particular why would these two construction tools specifically be privileged? You probably want to be something like https://github.com/whitequark/solvespace/issues/117.

> 
> Fixed "PNG" file export to export image right side up with out the toolbar in view.

Fixed in master long ago.

> 
> Added menu entry "Save With Exports" command to automatically save a "png", "html" and "stl"
> file along side the "slvs" file.

That's way too much junk to save by default, not to mention it is much slower. So at least this needs some configuration UI.

**Comment by whitequark** [comment]
@mmiscool I've implemented a command-line interface in master, through which you can perform batch exports, of any axonometric views and with any filename etc. Hope that helps!
